lib/icingadb/icingadb.cpp

```diff
@@ -145,6 +171,8 @@ void IcingaDB::Start(bool runtimeCreated)
 
 	m_Rcon->SuppressQueryKind(Prio::CheckResult);
 	m_Rcon->SuppressQueryKind(Prio::RuntimeStateSync);
+
+	m_HistoryThread = std::async(std::launch::async, [this]() { ForwardHistoryEntries(); });
```

Comment:
All references to `this` won't be destroyed.

Reply:
Please elaborate.

Reply:
Took you long enough to ask. 😉 Srsly:

https://github.com/Icinga/icinga2/blob/6048d0e8004daf647828cd55dc64b4abdc3a83cf/lib/cli/daemoncommand.cpp#L552
https://github.com/Icinga/icinga2/blob/6048d0e8004daf647828cd55dc64b4abdc3a83cf/lib/cli/daemoncommand.cpp#L323
https://github.com/Icinga/icinga2/blob/6048d0e8004daf647828cd55dc64b4abdc3a83cf/lib/base/application.cpp#L1011
https://github.com/Icinga/icinga2/blob/6048d0e8004daf647828cd55dc64b4abdc3a83cf/lib/icinga/icingaapplication.cpp#L107
https://github.com/Icinga/icinga2/blob/6048d0e8004daf647828cd55dc64b4abdc3a83cf/lib/base/application.cpp#L353-L356
https://github.com/Icinga/icinga2/blob/ee705bb110e802f8cafd21bab2d8697b0a538b0a/lib/icinga/icingaapplication.cpp#L121

There's no actual destruction, even here:

https://github.com/Icinga/icinga2/blob/ee705bb110e802f8cafd21bab2d8697b0a538b0a/lib/base/configobject.cpp#L619

Reply:
Indeed, looks like a normal shutdown does not destruct any config objects. But can we please use the obviously safe option rather than the one where we have to argue across the code base whether it's safe?

And then there are surprise features:

```diff
diff --git a/lib/icingadb/icingadb.hpp b/lib/icingadb/icingadb.hpp
index 785cd01ba..397edd37b 100644
--- a/lib/icingadb/icingadb.hpp
+++ b/lib/icingadb/icingadb.hpp
@@ -36,6 +36,8 @@ public:
 
        IcingaDB();
 
+       ~IcingaDB() { Log(LogCritical, "IcingaDB") << "~IcingaDB()"; }
+
        static void ConfigStaticInitialize();
        static void StatsFunc(const Dictionary::Ptr& status, const Array::Ptr& perfdata);
 
```

```
[2022-03-01 13:58:44 +0100] information/ApiListener: New client connection from [::ffff:172.31.0.1]:60864 (no client certificate)
[2022-03-01 13:58:44 +0100] information/IcingaDB: 'test' started.
[2022-03-01 13:58:44 +0100] information/IcingaDB: Trying to connect to Redis server (async) on host '127.0.0.1:6380' 
[2022-03-01 13:58:44 +0100] critical/IcingaDB: Cannot connect to 127.0.0.1:6380: Connection refused
[2022-03-01 13:58:44 +0100] information/ConfigObjectUtility: Created and activated object 'test' of type 'IcingaDB'.
[2022-03-01 13:58:44 +0100] information/HttpServerConnection: Request: PUT /v1/objects/icingadbs/test?pretty=1 (from [::ffff:172.31.0.1]:60864), user: root, agent: curl/7.81.0, status: OK).
[2022-03-01 13:58:44 +0100] information/HttpServerConnection: HTTP client disconnected (from [::ffff:172.31.0.1]:60864)
[...]
[2022-03-01 13:58:47 +0100] information/ApiListener: New client connection from [::ffff:172.31.0.1]:60866 (no client certificate)
[2022-03-01 13:58:47 +0100] information/IcingaDB: Flushing history data buffer to Redis. 
[2022-03-01 13:58:48 +0100] critical/IcingaDB: history: 3 queries failed (attempt #3) while we're about to shut down. Giving up and discarding additional 3 queued history queries.
[2022-03-01 13:58:48 +0100] information/IcingaDB: 'test' stopped.
[2022-03-01 13:58:48 +0100] critical/IcingaDB: Error during sending query 'HDEL' 'icinga:icingadb' '7328d7414728d34f4f1b1edc9953968d65d38b9f' which has been fired and forgotten:  
[2022-03-01 13:58:48 +0100] information/ConfigObjectUtility: Deleted object 'test' of type 'IcingaDB'.
[2022-03-01 13:58:48 +0100] critical/IcingaDB: ~IcingaDB()
[2022-03-01 13:58:48 +0100] information/HttpServerConnection: Request: DELETE /v1/objects/icingadbs/test?pretty=1 (from [::ffff:172.31.0.1]:60866), user: root, agent: curl/7.81.0, status: OK).
[2022-03-01 13:58:48 +0100] information/HttpServerConnection: HTTP client disconnected (from [::ffff:172.31.0.1]:60866)
```